tree/ntuple/v7/src/RClusterPool.cxx

```diff
@@ -0,0 +1,325 @@
+
+   // Determine previous cluster ids that we keep if they happen to be in the pool
+   std::set<DescriptorId_t> keep;
+   auto prev = clusterId;
+   for (unsigned int i = 0; i < fWindowPre; ++i) {
+      prev = desc.FindPrevClusterId(prev);
+      if (prev == kInvalidDescriptorId)
+         break;
+      keep.insert(prev);
+   }
+
+   // Determine following cluster ids and the column ids that we want to make available
+   RProvides provide;
+   provide.Insert(clusterId, columns);
+   auto next = clusterId;
+   for (unsigned int i = 1; i < fWindowPost; ++i) {
+      next = desc.FindNextClusterId(next);
+      if (next == kInvalidDescriptorId)
+         break;
+      provide.Insert(next, columns);
+   }
+
+   // Clear the cache from clusters not the in the look-ahead or the look-back window
+   for (auto &cptr : fPool) {
+      if (!cptr)
+         continue;
+      if (provide.Contains(cptr->GetId()) > 0)
+         continue;
+      if (keep.count(cptr->GetId()) > 0)
+         continue;
+      cptr.reset();
+   }
+
+   // Move clusters that meanwhile arrived into cache pool
+   {
+      std::lock_guard<std::mutex> lockGuardInFlightClusters(fLockInFlightClusters);
+      for (auto itr = fInFlightClusters.begin(); itr != fInFlightClusters.end(); ) {
+         R__ASSERT(itr->fFuture.valid());
+         itr->fIsExpired = !provide.Contains(itr->fClusterId) && (keep.count(itr->fClusterId) == 0);
+
+         if (itr->fFuture.wait_for(std::chrono::seconds(0)) != std::future_status::ready) {
+            // Remove the set of columns that are already scheduled for being loaded
+            provide.Erase(itr->fClusterId, itr->fColumns);
+            ++itr;
+            continue;
+         }
+
+         auto cptr = itr->fFuture.get();
+         // If cptr is nullptr, the cluster expired previously and was released by the I/O thread
+         if (!cptr || itr->fIsExpired) {
+            cptr.reset();
+            itr = fInFlightClusters.erase(itr);
+            continue;
+         }
+
+         // We either put a fresh cluster into a free slot or we merge the cluster with an existing one
+         auto existingCluster = FindInPool(cptr->GetId());
+         if (existingCluster) {
+            existingCluster->MergeCluster(std::move(*cptr));
+         } else {
+            auto idxFreeSlot = FindFreeSlot();
+            fPool[idxFreeSlot] = std::move(cptr);
+         }
+         itr = fInFlightClusters.erase(itr);
+      }
+
+      // Deterime clusters which get triggered for background loading
+      for (auto &cptr : fPool) {
+         if (!cptr)
+            continue;
+         provide.Erase(cptr->GetId(), cptr->GetAvailColumns());
+      }
+
+      // Update the work queue and the in-flight cluster list with new requests
+      // TODO(jblomer): we should ensure that clusterId is given first to the I/O thread.  That is usually the
+      // case but it's not ensured by the code
+      std::unique_lock<std::mutex> lockWorkQueue(fLockWorkQueue);
```

Comment:
This seems to be a nested lock.  Is there any risk of the 2 locks being taken in the other order?

Reply:
They are nowhere else taken in reversed order (actually nowhere else taken together). I added a clarifying comment.

Reply:
Even indirectly (i.e. a routine taking the 2nd lock then calling another routine that eventually takes the 1st one?)

Reply:
I re-checked all the places where the locks are taken, as far as I can see that's all fine.

Reply:
... eventually merged the two mutex